### PR TITLE
[FLINK-7041] Deserialize StateBackend from JobCheckpointingSettings with user classloader

### PR DESCRIPTION
**Pull request by aljoscha**
R: @tzulitai 

CC: @StephanEwen I noticed that the state backend that is in the configuration is never used anywhere. I'm guessing this is for future work or is something missing in the code?

### CODE REVIEW

**Comment by tzulitai** [comment]
Thanks for the quick fix @aljoscha!

Regarding tests:
Can we also have another test that verifies `JobGraphs` containing custom `StateBackend`s or `StateBackends` with custom classes can be submitted without exceptions?
Looks like `org.apache.flink.runtime.jobmanager.JobSubmitTest` is a good place for that.

**Comment by tzulitai** [comment]
Ah scratch that, perhaps the `CheckpointSettingsSerializableTest` is sufficient.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java [hunk 1]
```diff
@@ -223,19 +223,21 @@ public static ExecutionGraph buildGraph(
...
 			if (applicationConfiguredBackend != null) {
-				metadataBackend = applicationConfiguredBackend;
+				try {
+					metadataBackend = applicationConfiguredBackend.deserializeValue(classLoader);
+				} catch (IOException | ClassNotFoundException e) {
+					throw new JobExecutionException(jobId, "Could not instantiate configured state backend.", e);
+				}
 
 				log.info("Using application-defined state backend for checkpoint/savepoint metadata: {}.",
 						applicationConfiguredBackend);
```
Comment (tzulitai) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java]: Need to change the argument for this log to `metadataBackend`

Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraphBuilder.java]: Good catch!

**Comment by tzulitai** [comment]
Thanks for the fixups @aljoscha :-D
The last Travis run was successful, the test name rename should not affect anything.
Merging this ..

**Comment by aljoscha** [comment]
Thanks! 😃

**Pull request closed without merging**